---
title: "Machine learning for biological contexts"
output: 
  html_notebook:
    toc: true
    toc_float: true
date: 2021
---

_This material has been adapted from [ALSF CCDL training materials](https://github.com/AlexsLemonade/training-modules/tree/master/machine-learning)._

As we've seen in the course so far, we can explore data with unsupervised machine learning approaches like clustering or PCA.
Often, these methods can work with any generic dataset. 
In this notebook, we'll introduce a machine learning technique that is specifically for gene expression data.

The dataset we're using comes from the [OpenPBTA project](https://github.com/AlexsLemonade/OpenPBTA-analysis). 
We'll be using medulloblastoma data only.

## Set Up

```{r}
# Bit o' data wranglin'
library(tidyverse)
```

Make sure `PLIER` is installed.

```{r}
if (!("remotes" %in% installed.packages())) {
  install.packages("remotes")
}

if (!("PLIER" %in% installed.packages())) {
  # Install PLIER from GitHub
  remotes::install_github("wgmao/PLIER")
}
  
# Load Pathway-Level Information ExtractoR
library(PLIER)
```

```{r set_seed}
# We will need this for our dimension reduction to be reproducible and there is
# some sampling that goes into collapsing our RNA-seq matrix
seed <- 1234
set.seed(seed)
```

## Read in and set up data

### RNA-seq data

```{r}
# Read in transformed RNA-seq data
rnaseq_df <- read_tsv(file.path("data", 
                                "expression", 
                                "pbta-vst-medulloblastoma.tsv.gz"))
```
Let's take a look at what's in the `rnaseq_df`.

```{r}
head(rnaseq_df)
```

The `gene_id` column concatenates two different gene identifiers: Ensembl gene IDs and gene symbols.
PLIER, which we will use later in this notebook, requires gene symbols as rownames, so we will need to take some steps to prepare this data for use with PLIER.

First, we need to separate the gene symbols from the Ensembl gene identifiers.
Luckily, there's a tidyverse function that can separate one column into multiple columns called `tidyr::separate()`.

```{r separate_gene_ids}
symbol_rnaseq_df <- rnaseq_df %>%
  tidyr::separate(gene_id,  # Take the gene_id column
                  # Create two new columns called ensembl_id and gene_symbol
                  into = c("ensembl_id", "gene_symbol"),
                  # The values for these two columns are separated by _
                  sep = "_",
                  # Some gene symbols themselves contain _, so when that happens
                  # merge everything after the first _ into the gene_symbol
                  # column
                  extra = "merge")


# We no longer need the data frame we read in, so we'll remove it to save some
# memory
rm(rnaseq_df)
```

Alright, what does our new data frame look like?

```{r head_symbol_df, live = TRUE}
head(symbol_rnaseq_df)
```

We won't need the column with the Ensembl identifiers anymore, so let's remove it.

```{r remove_ensg, live = TRUE}
symbol_rnaseq_df <- symbol_rnaseq_df %>%
  select(-ensembl_id)
```

### Collapsing duplicate gene symbols

```{r any_duplicated}
any(duplicated(symbol_rnaseq_df$gene_symbol))
```

How many duplicated gene symbols are there?

```{r how_many_duplicated}
sum(duplicated(symbol_rnaseq_df$gene_symbol))
```

Relative to the total number of genes, there are not too many genes that are duplicated.

```{r add_row_means}
# mean transformed value across the entire cohort
symbol_rnaseq_df$mean_value <- rowMeans(symbol_rnaseq_df[, -1])
```

We'll use a few gene symbols that we've arbitrarily selected to follow along with these processing steps!

```{r duplicated_examples}
# Genes we will use to illustrate what happens during the collapsing duplicate
# process
duplicated_symbols <- c("DGCR5", "RABGEF1")

# Let's take a look at the mean values prior to any filtering
symbol_rnaseq_df %>% 
  filter(gene_symbol %in% duplicated_symbols) %>%
  select(gene_symbol, mean_value) %>%
  arrange(gene_symbol)
```

For a given duplicated gene symbol, we want to select the row with the highest average expression value.

```{r collapsed_rnaseq}
collapsed_rnaseq_df <- symbol_rnaseq_df %>%
  # For each set of rows that correspond to the same gene symbol
  group_by(gene_symbol) %>%
  # Select the single row with the highest value in the mean_value column
  top_n(1, mean_value) %>%
  # In the case of ties, where the mean values are the same, randomly pick one
  # row
  sample_n(1)
```

Check again to see if any duplicated gene symbols remain in the new collapsed data frame.

```{r any_duplicated_now, live = TRUE}
any(duplicated(collapsed_rnaseq_df$gene_symbol))
```

Which row for our gene symbols of interest was retained in the collapsed data frame?

```{r which_retained}
collapsed_rnaseq_df %>% 
  filter(gene_symbol %in% duplicated_symbols) %>%
  select(gene_symbol, mean_value) %>%
  arrange(gene_symbol)
```

Make a matrix.

```{r get_medulloblastoma_mat, live = TRUE}
medulloblastoma_rnaseq_mat <- collapsed_rnaseq_df %>%
  select(-mean_value) %>%
  tibble::column_to_rownames("gene_symbol") %>%
  as.matrix()
```

And a bit of cleanup.

```{r remove_old_df}
rm(collapsed_rnaseq_df, symbol_rnaseq_df)
```


## Pathway-Level Information ExtractoR (PLIER)

In this notebook, we'll use a method called Pathway-Level Information Extractor (PLIER) ([Mao *et al.* (2019)](https://doi.org/10.1038/s41592-019-0456-1)).

We like PLIER for a few reasons:

* It is a matrix factorization approach. 
  That means we can get a low-dimensional representation of our data.
  Specifically, PLIER learns correlated patterns of expression in our data or latent variables (LVs).
  Here, a latent variable is an "eigengene-like" combination of genes' expression.
  (It's called _latent_ because it's not directly measured, but instead inferred from the individual gene measurements.)
* It includes penalties such that _some_ of the LVs will align with gene sets that we give it, so it's excellent for biological discovery.
* The authors demonstrated that it performs favorably with regard to estimating proportion of immune cells in a sample as compared to other methods.
* Because not _all_ LVs align with the gene sets we input, some of them capture unwanted technical variation. 
  In our experience with the method, it does this quite well.

PLIER is similar to other pathway analysis methods that you may be familiar with in that it uses prior knowledge in the form of gene sets.
It produces output values that are on an individual sample level and does not require a two group comparison ahead of time like some pathway analysis methods.
However, PLIER is designed to align the LVs it constructs with the relevant input gene sets that the data supports, whereas other methods will use all gene sets you provide as input.

Here's an overview of the PLIER method from [Mao _et al._ (2019)](https://doi.org/10.1038/s41592-019-0456-1) (Figure 1).

![](diagrams/mao_nature_methods_fig1.png)

> **Fig. 1 | PLIER overview.** 
PLIER is a matrix factorization approach that decomposes gene expression data into a product of a small number of LVs and their corresponding gene associations or loadings, while constraining the loadings to align with the most relevant automatically selected subset of prior knowledge. **a**, Given two inputs, the gene expression matrix _Y_ and the prior knowledge (represented as binary gene set membership in matrix _C_), the method returns the LVs (_B_), their loadings (_Z_), and an additional sparse matrix (_U_) that specifies which (if any) prior-information gene sets and pathways are used for each LV. The light gray area of _U_ indicates the large number of zero elements of the matrix. We apply our method to a whole-blood human gene expression dataset. **b**, The positive entries of the resulting _U_ matrix are visualized as a heat map, facilitating the identification of the correspondence between specific LVs and prior biological knowledge. As the absolute scale of the _U_ matrix is arbitrary, each column is normalized to a maximum of 1. **c**, We validate the LVs mapped to specific leukocyte cell types by comparing PLIER estimated relative cell-type proportions with direct measurements by mass cytometry. Dashed lines represent 0.05, 0.01, and 0.001 significance levels for Spearman rank correlation (one-tailed test). NK cell, natural killer cell.

### Pathway data

The `PLIER` package comes with different gene sets we can use as input during training.
The gene sets themselves are not really any different than other gene sets we might use for pathway analysis, but the data that comes in the package is in a format that is already compatible for use with PLIER.

```{r plier_data}
data("bloodCellMarkersIRISDMAP")
data("canonicalPathways")
data("svmMarkers")

# Combine the pathway data from PLIER
all_pathways <- PLIER::combinePaths(bloodCellMarkersIRISDMAP, 
                                    canonicalPathways,
                                    svmMarkers)
```

### Row normalize

`PLIER` comes with a built-in function to row normalize the expression data.
This is z-scoring the expression values for each gene.

```{r row_normalize, live = TRUE}
# Row normalize -- z-scoring the genes
medulloblastoma_zscore <- PLIER::rowNorm(medulloblastoma_rnaseq_mat)
```

### Run PLIER

```{r common_genes}
# Identify genes that are common to the pathway data and the zscore mat
common_genes <- PLIER::commonRows(all_pathways, medulloblastoma_zscore)
```

Now for the main PLIER function.
This will take a bit to run.

```{r run_plier}
plier_results <- PLIER::PLIER(data = medulloblastoma_zscore[common_genes, ],
                              priorMat = all_pathways[common_genes, ],
                              rseed = seed)
```

### Explore the model

What does the output of `PLIER::PLIER()` look like?

```{r view_plier, live = TRUE, eval = FALSE}
View(plier_results)
```

The _U_ matrix tells us about how the latent variables learned by the model relate to the pathways we used as input.
`plotU()` is a special function to display the _U_ matrix.

```{r plot_u}
PLIER::plotU(plier_results,
             fontsize_row = 6)
```

`summary()` of a `PLIER` results object returns the FDR and AUC values for input pathway to latent variable relationships.

```{r lv_summary}
plier_results$summary %>% 
  filter(FDR < 0.05) %>%
  arrange(FDR)
```

The _B_ matrix contains the latent variable values for each sample.

```{r dim_b}
dim(plier_results$B)
```

Let's take a peek at the matrix itself.

```{r b_preview, live = TRUE}
plier_results$B[1:5, 1:5]
```

The _Z_ matrix contains the gene loadings (how genes combine to get _B_).

```{r dim_z}
dim(plier_results$Z)
```

We can use _Z_ to tell us which genes contribute to individual LVs by accessing the column corresponding to that LV.
We'll use 20 below, but you can change the number to suit your purposes!

```{r lv_loadings}
head(sort(plier_results$Z[, 20], 
          decreasing = TRUE), 
     n = 25)
```

## Biological exploration with PLIER

### First, a note on tidy data

In order to use `ggplot2`, we'll need the data in "long" or "tidy" format. 
`PLIER` outputs what we want to plot in what we call "wide" format.

Read more about tidy data [here](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html).

To quote from Hadley Wickham's [R for Data Science](https://r4ds.had.co.nz/):

> There are three interrelated rules which make a dataset tidy:
>
> * Each variable must have its own column.
>
> * Each observation must have its own row.
>
> * Each value must have its own cell.

Let's look at a toy example.

```{r create_toy_wide}
set.seed(12345)
toy_df <- data.frame( 
  cbind(c("GENEA", "GENEB", "GENEC"),
        matrix(rnorm(30), ncol = 10))
)
colnames(toy_df) <- c("Gene", paste0("Sample", 1:10))
```

`toy_df` is now in "wide" format.

```{r toy_df}
toy_df
```

Let's get it into "long" format.

```{r make_toy_long}
toy_long_df <- tidyr::pivot_longer(toy_df,
                                   # The data is in every column except the one 
                                   # named "Gene"
                                   cols = -Gene, 
                                   # What will we call the column of the old df 
                                   # column names?
                                   names_to = "Sample",
                                   # What will we call the column of values 
                                   # from the old df?
                                   values_to = "Expression")

toy_long_df
```

Let's remove these toy examples from the workspace.

```{r remove_toy, live = TRUE}
rm(toy_df, toy_long_df)
```

### LVs associated with pathways

For biological discovery, we are often most interested in the latent variables that have some kind of association with an input gene set or pathway.
We can use the FDR values in the summary data frame to filter to only the latent variables with a significant association (and their associated gene sets).

```{r filter_fdr, live = TRUE}
# Filter to LV-pathway relationships with FDR < 0.05
sig_summary_df <- plier_results$summary %>%
  dplyr::filter(FDR < 0.05)
sig_summary_df
```

```{r sig_index}
# We only want a single instance of each LV index
sig_index <- as.integer(unique(sig_summary_df$`LV index`))
```

```{r sig_b_wide, live = TRUE}
# Get the LV by sample matrix from the PLIER results and subset it to only those
# LVs with an FDR < 0.05 (at least one pathway)
b_matrix <- plier_results$B
sig_b_matrix <- b_matrix[sig_index, ]

# Create a data frame and add a column with LV identifiers
sig_b_wide <- data.frame(sig_b_matrix) %>%
  tibble::rownames_to_column(var = "LV")
```

#### Tidy

Let's look at what format this is in currently.

```{r wide_peek}
sig_b_wide
```

We want this in long format for plotting.
We'll use `tidyr::pivot_longer` to do this just like in the toy example above.

```{r pivot_longer_b, live = TRUE}
sig_b_df <- tidyr::pivot_longer(sig_b_wide,
                                cols = starts_with("BS_"),
                                names_to = "Kids_First_Biospecimen_ID",
                                values_to = "LV_estimate")
head(sig_b_df)
```

Right now the `LV` column has values that contain two pieces of information: the LV index and the pathway that the LV has been named for.

*Remember, just because a LV is named for a single pathway, that doesn't mean that that is the only input pathway that is significantly associated with that latent variable - always check `summary`!*

Now let's add relevant metadata to the data frame so we can use that for plotting.
We'll need to read it in first!

```{r add_metadata_b}
# Read in metadata
histologies_df <- read_tsv(file.path("data", 
                                     "metadata", 
                                     "pbta-histologies-stranded-rnaseq.tsv"))

# Create a data frame that only has the biospecimen identifiers and the 
# molecular subtype labels
subtype_df <- histologies_df %>%
  select(Kids_First_Biospecimen_ID,
         molecular_subtype)

# Add the subtype labels to the LV estimates
sig_b_df <- inner_join(x = sig_b_df, 
                       y = subtype_df,
                       by = "Kids_First_Biospecimen_ID")
```
## Plotting

We'll plot LV20; this is the latent variable that we looked at the loadings for in an earlier chunk.

```{r lv_to_plot}
# PLIER names certain latent variables based on their association with input
# gene sets
lv_to_plot <- rownames(plier_results$B)[20]
# For plotting, subset only to the rows corresponding to this latent variable
lv_plot_df <- sig_b_df %>%
  filter(LV == lv_to_plot) 
```

### Boxplot and customization

Let's start by making a simple boxplot.

```{r lv_boxplot, live = TRUE}
# Make a boxplot where samples are grouped by molecular subtype
ggplot(lv_plot_df,
       aes(x = molecular_subtype, 
           y = LV_estimate, 
           group = molecular_subtype, 
           colour = molecular_subtype)) +
  geom_boxplot()
```

It can often be helpful to visualize individual samples.

```{r lv_jitter, live = TRUE}
# Add individual points with geom_jitter()
ggplot(lv_plot_df,
       aes(x = molecular_subtype, 
           y = LV_estimate, 
           group = molecular_subtype, 
           colour = molecular_subtype)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter()
```

We're able to globally adjust the aesthetics of the jitter points.

```{r improve_jitter, live = TRUE}
# Improve the aesthetics of the points
ggplot(lv_plot_df,
       aes(x = molecular_subtype, 
           y = LV_estimate, 
           group = molecular_subtype, 
           colour = molecular_subtype)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.5)
```

Add a built-in `ggplot2` theme.

```{r theme_bw, live = TRUE}
# Use @jaclyn-taroni's favorite theme :)
ggplot(lv_plot_df,
       aes(x = molecular_subtype, 
           y = LV_estimate, 
           group = molecular_subtype, 
           colour = molecular_subtype)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.5) +
  theme_bw()
```

Use the same color palette we used in the UMAP plot earlier in the module.

```{r okabe_ito, live = TRUE}
# Add "Okabe-Ito" color scheme
ggplot(lv_plot_df,
       aes(x = molecular_subtype, 
           y = LV_estimate, 
           group = molecular_subtype, 
           colour = molecular_subtype)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.5) +
  theme_bw() +
  scale_color_manual(values = unname(palette.colors(palette = "Okabe-Ito")))
```

Add a title to the plot.

```{r add_title, live = TRUE}
# Use labs() to add a title
ggplot(lv_plot_df,
       aes(x = molecular_subtype, 
           y = LV_estimate, 
           group = molecular_subtype, 
           colour = molecular_subtype)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.5) +
  theme_bw() +
  scale_color_manual(values = unname(palette.colors(palette = "Okabe-Ito"))) +
  labs(title = lv_to_plot)
```

Center the title and make it bigger and in bold.

```{r center_title, live = TRUE}
# Use theme() to improve the way the title looks
ggplot(lv_plot_df,
       aes(x = molecular_subtype, 
           y = LV_estimate, 
           group = molecular_subtype, 
           colour = molecular_subtype)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.5) +
  theme_bw() +
  scale_color_manual(values = unname(palette.colors(palette = "Okabe-Ito"))) +
  labs(title = lv_to_plot) +
  theme(plot.title = element_text(size = 15,
                                  face = "bold",
                                  hjust = 0.5))
```

Use the next chunks to further customize your plot.
We might suggest starting with [the x- and y-axis labels](https://ggplot2.tidyverse.org/reference/labs.html).

```{r labs_help, eval = FALSE}
?labs
```

```{r adventure_time}

```

## Session Info

```{r}
sessionInfo()
```

